### PR TITLE
[FIX](decimalv3) fix decimalv3 with precision cast 

### PR DESCRIPTION
**Pull request by amorynan**
## Proposed changes

Issue Number: close #xxx
now we use cast to decimalv3 may has error , because decimalv3 use type precision for translate string
```
mysql [test]>select cast("9999e-1" as decimal(2,1));
+------------------------------------+
| cast('9999e-1' as DECIMALV3(2, 1)) |
+------------------------------------+
|                              999.9 |
+------------------------------------+
1 row in set (0.01 sec)
```
this pr will fix this just keep reaction same with mysql
```
mysql> select cast('9999e-1' as decimalv3(2, 1));
+------------------------------------+
| cast('9999e-1' as DECIMALV3(2, 1)) |
+------------------------------------+
|                                9.9 |
+------------------------------------+
1 row in set (0.07 sec)
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by amorynan** [comment]
run buildall

File: be/src/util/string_parser.hpp [hunk 1]
```diff
@@ -714,9 +716,9 @@ T StringParser::string_to_decimal(const char* s, int len, int type_precision, in
                     return 0;
                 }
                 *result = StringParser::PARSE_SUCCESS;
-                value *= get_scale_multiplier<T>(type_scale - scale);
-
-                return is_negative ? T(-value) : T(value);
+                if (type_scale > scale) {
+                    value *= get_scale_multiplier<T>(type_scale - scale);
```
Comment (yiguolei) [inline be/src/util/string_parser.hpp:720]: missed return???

File: be/src/util/string_parser.hpp [hunk 2]
```diff
@@ -746,6 +748,14 @@ T StringParser::string_to_decimal(const char* s, int len, int type_precision, in
     *result = StringParser::PARSE_SUCCESS;
     if (UNLIKELY(precision - scale > type_precision - type_scale)) {
         *result = StringParser::PARSE_OVERFLOW;
+        if constexpr (TYPE_DECIMALV2 == P) {
```
Comment (yiguolei) [inline be/src/util/string_parser.hpp:751]: clang tidy check will fail

File: be/src/util/string_parser.hpp (hunk 1 above)
Comment (amorynan) [reply be/src/util/string_parser.hpp:720]: we should make loop to meet E key word

File: be/src/vec/data_types/data_type_decimal.h [hunk 3]
```diff
@@ -387,9 +387,9 @@ constexpr bool IsDataTypeDecimalOrNumber =
 
 template <typename FromDataType, typename ToDataType>
     requires IsDataTypeDecimal<FromDataType> && IsDataTypeDecimal<ToDataType>
-ToDataType::FieldType convert_decimals(const typename FromDataType::FieldType& value,
-                                       UInt32 scale_from, UInt32 scale_to,
-                                       UInt8* overflow_flag = nullptr) {
+typename ToDataType::FieldType convert_decimals(const typename FromDataType::FieldType& value,
```
Comment (yiguolei) [inline be/src/vec/data_types/data_type_decimal.h:390]: why add this??

Comment (amorynan) [reply be/src/vec/data_types/data_type_decimal.h:390]: clion compiler auto optimize , I will drop this

**Comment by amorynan** [comment]
run clickbench

**Review by Gabriel39 — APPROVED** [review]
LGTM

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**